src/huntsman/pocs/camera/pyro/client.py

```diff
@@ -274,20 +284,32 @@ def autofocus(self, blocking=False, timeout=None, coarse=False, *args, **kwargs)
 
         return self._focus_event
 
+    def process_exposure(self, *args, **kwargs):
+        """ Small wrapper around `Camera.process_exposure` that makes sure the image is actually
+        written before starting processing. """
+        self._wait_for_file_thread.join()
+        return super().process_exposure(*args, **kwargs)
+
     # Private Methods
-    def _wait_for_readout(self, timeout, sleep_time=0.1):
-        """ Wait for readout to finish on the camera service.
+    def _wait_for_file(self, filename, timeout, sleep_time=0.1):
+        """ Wait for the file to be written.
         Args:
-            timeout (float): The readout timeout in seconds.
+            timeout (float): The timeout in seconds.
         """
         proxy = self._proxy
         timer = CountdownTimer(timeout)
         while not timer.expired():
             if not proxy.is_reading_out:
-                return
+                try:
+                    # The best way of checking the file is written appears to be to get its data
+                    fits.getdata(filename)
```

Comment:
This does seem like an expensive way of checking for a valid file, especially in a loop. The data is also not accessed (it's memory mapped) by default, so not sure this is actually doing what you would want.

What is the error this is trying to solve? As in, where does the problem arise if the file is not valid?

Reply:
@wtgee The problem is that there is a delay between the camera service saying it has written the image and it actually appearing as a complete FITS file on the control computer. So sometimes we try and read a partial FITS file. From my experience, the header can actually be read from an incomplete FITS file - calling `getadata` is (again from my experience) the best way of checking if the file is fully written because it uses the FITS `CHECKSUM`. If it is memory mapped then it is a relatively cheap operation, I think.